README.md
# 🧮 bloom-filter-uuid

A Bloom filter for UUIDs.

## 🎯 Motivation

This filter is a variant of a Bloom filter optimized for work with UUIDs.
It has variants for various desired amounts of data and false positive probabilities.

## ✨ Features

- Easy in use
- Optimized for UUIDs
- Able to allocate arbitrary huge amount of memory
- Space-efficient
- Clear and predictable behavior

## 📝 Usage

```js
const BloomFilter = require('@vadym.gidulian/bloom-filter-uuid');

// Print help
BloomFilter.printHelp();
/*
A Bloom filter is a space-efficient probabilistic data structure, that is used to test whether an element is a member of a set.
False positive matches are possible, but false negatives are not – in other words, a query returns either "possibly in set" or "definitely not in set".
Elements can be added to the set, but not removed; the more items added, the larger the probability of false positives.

                                   Max. number of inserted elements for desired false positive probability                                   
Variant |   Memory   |  0.001%  |   0.01%  |   0.1%   |    1%    |    3%    |    5%    |    10%   |    15%   |    20%   |    25%   |    50%  
--------+------------+----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+---------
   40   | 384    GiB |  23.9  B |  52.3  B | 116    B | 267    B | 409    B | 505    B | 686    B | 833    B | 966    B |   1.09 T |   1.74 T
   39   | 192    GiB |  12    B |  26.1  B |  57.9  B | 133    B | 205    B | 253    B | 343    B | 417    B | 483    B | 547    B | 868    B
   38   |  96    GiB |   5.99 B |  13.1  B |  29    B |  66.7  B | 102    B | 126    B | 172    B | 208    B | 242    B | 273    B | 434    B
   37   |  48    GiB |   2.99 B |   6.53 B |  14.5  B |  33.3  B |  51.1  B |  63.2  B |  85.8  B | 104    B | 121    B | 137    B | 217    B
   36   |  24    GiB |   1.5  B |   3.27 B |   7.24 B |  16.7  B |  25.6  B |  31.6  B |  42.9  B |  52.1  B |  60.4  B |  68.3  B | 108    B
   35   |  12    GiB | 748    M |   1.63 B |   3.62 B |   8.34 B |  12.8  B |  15.8  B |  21.4  B |  26    B |  30.2  B |  34.2  B |  54.2  B
   34   |   6    GiB | 374    M | 817    M |   1.81 B |   4.17 B |   6.39 B |   7.89 B |  10.7  B |  13    B |  15.1  B |  17.1  B |  27.1  B
   33   |   3    GiB | 187    M | 408    M | 905    M |   2.08 B |   3.2  B |   3.95 B |   5.36 B |   6.51 B |   7.55 B |   8.54 B |  13.6  B
   32   |   2    GiB | 249    M | 453    M | 841    M |   1.63 B |   2.31 B |   2.75 B |   3.55 B |   4.18 B |   4.75 B |   5.27 B |   7.9  B
   31   |   1    GiB | 124    M | 226    M | 420    M | 816    M |   1.16 B |   1.38 B |   1.77 B |   2.09 B |   2.37 B |   2.64 B |   3.95 B
   30   | 512    MiB |  62.1  M | 113    M | 210    M | 408    M | 578    M | 688    M | 887    M |   1.05 B |   1.19 B |   1.32 B |   1.97 B
   29   | 256    MiB |  31.1  M |  56.6  M | 105    M | 204    M | 289    M | 344    M | 444    M | 523    M | 593    M | 659    M | 987    M
   28   | 128    MiB |  15.5  M |  28.3  M |  52.6  M | 102    M | 144    M | 172    M | 222    M | 261    M | 297    M | 330    M | 493    M
   27   |  64    MiB |   7.77 M |  14.1  M |  26.3  M |  51    M |  72.2  M |  85.9  M | 111    M | 131    M | 148    M | 165    M | 247    M
   26   |  32    MiB |   3.88 M |   7.07 M |  13.1  M |  25.5  M |  36.1  M |  43    M |  55.5  M |  65.3  M |  74.1  M |  82.4  M | 123    M
   25   |  20    MiB |   3.54 M |   5.79 M |   9.71 M |  17    M |  23    M |  26.7  M |  33.4  M |  38.7  M |  43.3  M |  47.6  M |  68.6  M
   24   |  10    MiB |   1.77 M |   2.9  M |   4.85 M |   8.52 M |  11.5  M |  13.4  M |  16.7  M |  19.3  M |  21.6  M |  23.8  M |  34.3  M
   23   |   5    MiB | 884    K |   1.45 M |   2.43 M |   4.26 M |   5.75 M |   6.68 M |   8.36 M |   9.67 M |  10.8  M |  11.9  M |  17.2  M
   22   |   2.5  MiB | 442    K | 724    K |   1.21 M |   2.13 M |   2.87 M |   3.34 M |   4.18 M |   4.84 M |   5.41 M |   5.95 M |   8.58 M
   21   |   1.5  MiB | 333    K | 509    K | 797    K |   1.31 M |   1.71 M |   1.96 M |   2.4  M |   2.74 M |   3.03 M |   3.31 M |   4.65 M
   20   | 768    KiB | 166    K | 254    K | 399    K | 654    K | 855    K | 979    K |   1.2  M |   1.37 M |   1.52 M |   1.66 M |   2.32 M
   19   | 384    KiB |  83.2  K | 127    K | 199    K | 327    K | 427    K | 490    K | 600    K | 684    K | 759    K | 828    K |   1.16 M
   18   | 224    KiB |  56.2  K |  81.9  K | 122    K | 191    K | 244    K | 277    K | 333    K | 377    K | 415    K | 450    K | 619    K
   17   | 112    KiB |  28.1  K |  40.9  K |  61.1  K |  95.6  K | 122    K | 138    K | 167    K | 188    K | 207    K | 225    K | 310    K
   16   |  64    KiB |  17.7  K |  24.9  K |  35.9  K |  54.2  K |  67.9  K |  76.3  K |  90.8  K | 102    K | 112    K | 120    K | 163    K
   15   |  32    KiB |   8.87 K |  12.5  K |  17.9  K |  27.1  K |  33.9  K |  38.1  K |  45.4  K |  51    K |  55.8  K |  60.2  K |  81.6  K
   14   |  18    KiB |   5.34 K |   7.3  K |  10.2  K |  15    K |  18.5  K |  20.7  K |  24.4  K |  27.2  K |  29.6  K |  31.9  K |  42.6  K
   13   |   9    KiB |   2.67 K |   3.65 K |   5.11 K |   7.5  K |   9.27 K |  10.3  K |  12.2  K |  13.6  K |  14.8  K |  15.9  K |  21.3  K
   12   |   5    KiB |   1.56 K |   2.08 K |   2.85 K |   4.08 K |   4.99 K |   5.54 K |   6.48 K |   7.19 K |   7.81 K |   8.37 K |  11.1  K
   11   |   2.75 KiB | 886      |   1.16 K |   1.56 K |   2.2  K |   2.66 K |   2.94 K |   3.41 K |   3.77 K |   4.08 K |   4.37 K |   5.73 K
   10   |   1.5  KiB | 495      | 639      | 846      |   1.17 K |   1.41 K |   1.55 K |   1.79 K |   1.97 K |   2.13 K |   2.27 K |   2.95 K
    9   | 896      B | 296      | 374      | 483      | 651      | 772      | 843      | 966      |   1.06 K |   1.14 K |   1.21 K |   1.55 K
    8   | 512      B | 171      | 212      | 268      | 355      | 416      | 452      | 514      | 561      | 601      | 637      | 809     
    7   | 288      B |  96      | 117      | 146      | 191      | 222      | 240      | 271      | 295      | 315      | 333      | 419     
    6   | 168      B |  55      |  66      |  81      | 104      | 120      | 129      | 145      | 157      | 167      | 176      | 219     
    5   | 100      B |  32      |  38      |  45      |  57      |  65      |  70      |  78      |  84      |  89      |  93      | 115     
    4   |  64      B |  19      |  22      |  26      |  32      |  36      |  39      |  43      |  46      |  48      |  51      |  61     
    3   |  42      B |  11      |  13      |  15      |  18      |  20      |  21      |  23      |  25      |  26      |  27      |  33     
    2   |  32      B |   7      |   8      |   9      |  11      |  12      |  12      |  13      |  14      |  15      |  15      |  18     
*/

// Get information about a variant
const variant = BloomFilter.describeVariant(12);
/*
{
	k: 10,            // Number of hash functions
	m: 40960,         // Number of bits
	bufferSize: 5120, // Size in memory
	calcE, calcN
}
*/
// Calculate approximate false positive probability of the variant
// with specified desired number of inserted elements
variant.calcE(5_000); // 0.03032...
// Calculate approximate number of elements that could be inserted
// with specified desired false positive probability of the variant
variant.calcN(2_500); // 4083


const bf = new BloomFilter(12);
bf.variant; // 12
bf.buffers; // Array of buffers storing the set

// Add an element to the set
bf.add('01234567-89ab-cdef-0123-456789abcdef');
bf.add('...');
bf.add('...');

// Test whether an element is in the set
bf.has('01234567-89ab-cdef-0123-456789abcdef'); // true, with probability of `1 - variant.calcE(3)`
bf.has('fedcba98-7654-3210-fedc-ba9876543210'); // definitely false
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] &ndash; 2021-10-05


test/index.js
'use strict';

const childProcess = require('child_process');
const fs           = require('fs');
const path         = require('path')

const B   = 1;
const KiB = 2**10;
const MiB = 2**20;
const GiB = 2**30;

const DATA = [
	[35,  3,  12 * GiB],
	[34,  3,   6 * GiB],
	[33,  3,   3 * GiB],
	[32,  4,   2 * GiB],
	[31,  4,   1 * GiB],
	[30,  4, 512 * MiB],
	[29,  4, 256 * MiB],
	[28,  4, 128 * MiB],
	[27,  4,  64 * MiB],
	[26,  4,  32 * MiB],
	[25,  5,  20 * MiB],
	[24,  5,  10 * MiB],
	[23,  5,   5 * MiB],
	[22,  5,   2.5 * MiB],
	[21,  6,   1.5 * MiB],
	[20,  6, 768 * KiB],
	[19,  6, 384 * KiB],
	[18,  7, 224 * KiB],
	[17,  7, 112 * KiB],
	[16,  8,  64 * KiB],
	[15,  8,  32 * KiB],
	[14,  9,  18 * KiB],
	[13,  9,   9 * KiB],
	[12, 10,   5 * KiB],
	[11, 11,   2.75 * KiB],
	[10, 12,   1.5 * KiB],
	[ 9, 14, 896 * B],
	[ 8, 16, 512 * B],
	[ 7, 18, 288 * B],
	[ 6, 21, 168 * B],
	[ 5, 25, 100 * B],
	[ 4, 32,  64 * B],
	[ 3, 42,  42 * B],
	[ 2, 64,  32 * B]
];

console.info('Removing previous generated tests, if any...');
const generatedTestsDirPath = path.join(__dirname, 'generated');
fs.rmSync(generatedTestsDirPath, {force: true, recursive: true});
console.info('Removed.');


console.info('Generating tests...');
fs.mkdirSync(generatedTestsDirPath);

const template = fs.readFileSync(path.join(__dirname, '_template.js'), 'utf8');

for (const [variant, k, bufferSize] of DATA) {
	const fileContent = template
		.replace(/\$VARIANT\$/g,     variant)
		.replace(/\$K\$/g,           k)
		.replace(/\$BUFFER_SIZE\$/g, bufferSize);
	const filePath = path.join(generatedTestsDirPath, `variant-${variant}.test.js`);
	
	fs.writeFileSync(filePath, fileContent, 'utf8');
}
console.log('Generated.');


console.info('Executing...');
childProcess.spawnSync('npx', ['jest'], {stdio: 'inherit'});
console.info('Done.');


test/_template.js
'use strict';

const omit         = require('lodash.omit');
const {v4: uuidv4} = require('uuid');

const BloomFilter = require('../../src');

const variant    = $VARIANT$;
const k          = $K$;
const bufferSize = $BUFFER_SIZE$;
describe('Bloom filter (variant $VARIANT$)', () => {
	test('Variant description', () => {
		const variantDescription = BloomFilter.describeVariant(variant);
		expect(omit(variantDescription, ['calcE', 'calcN'])).toStrictEqual({k, m: bufferSize * 8, bufferSize});
		expect(typeof variantDescription.calcE).toBe('function');
		expect(typeof variantDescription.calcN).toBe('function');
	});
	
	test('Buffer size', () => {
		const bf = new BloomFilter(variant);
		
		expect(bf.buffers.reduce((sum, buffer) => sum + buffer.length, 0)).toStrictEqual(bufferSize);
	});
	
	{
		function getTestParams() {
			if      (variant <= 20) return [100, 0.01,       100_000];
			else if (variant <= 25) return [10,  0.001,    1_000_000];
			else if (variant <= 30) return [1,   0.0001,  10_000_000];
			else                    return [1,   0.00001, 10_000_000];
		}
		
		const [TRIES, e, N] = getTestParams();
		const {calcN}       = BloomFilter.describeVariant(variant);
		
		test(`False positive probability (${e*100}%) of a set with ${prettyNumber(calcN(e))} elements, ${TRIES} times`, () => {
			let falsePositives = 0;
			for (let i = 0; i < TRIES; i++) {
				const bf = new BloomFilter(variant);
				
				for (let j = 0; j < calcN(e); j++) {
					const uuid = uuidv4();
					bf.add(uuid);
					if (!bf.has(uuid)) throw new Error('False negative');
				}
				
				for (let j = 0; j < N; j++) {
					if (bf.has(uuidv4())) falsePositives++;
				}
			}
			
			const actualE = falsePositives / (TRIES * N);
			console.info(`Actual error: ${Number((actualE * 100).toFixed(6))}%`);
			if (variant === 2) {
				// Too inaccurate
				expect(actualE).toBeLessThan(0.075);
			} else if (variant === 3) {
				// Slightly inaccurate
				expect(actualE).toBeLessThan(0.025);
			} else {
				expect(actualE).toBeCloseTo(e, -Math.log10(e));
			}
		});
	}
});



function prettyNumber(value) {
	const n = Number(value.toPrecision(3));
	
	let divider = 1;
	let suffix  = '';
	for (const [d, s] of [[1e12, 'T'], [1e9, 'B'], [1e6, 'M'], [1e3, 'K']]) {
		if (n < d) continue;
		divider = d;
		suffix  = s;
		break;
	}
	
	const [ip, fp] = String(n / divider).split('.');
	return ip + (fp ? `.${fp}` : '') + suffix;
}


test/common.test.js
'use strict';

const BloomFilter = require('../src');

test.each([1, 0, -1, NaN, null, undefined])('Invalid variant description (%s)', (variant) => {
	expect(() => BloomFilter.describeVariant(variant)).toThrow(TypeError);
});


src/index.d.ts
interface VariantDescription {
	/**
	 * Number of hash functions
	 */
	k: number;
	/**
	 * Number of bits
	 */
	m: number;
	/**
	 * Size in memory
	 */
	bufferSize: number;
	/**
	 * Calculates approximate false positive probability of the variant
	 * with specified desired number of inserted elements.
	 * @param n - Number of elements
	 */
	calcE: (n: number) => number;
	/**
	 * Calculates approximate number of elements that could be inserted
	 * with specified desired false positive probability of the variant.
	 * @param e - Desired false positive probability
	 */
	calcN: (e: number) => number;
}

export default class BloomFilter {
	
	buffers: Buffer[];
	variant: number;
	
	/**
	 * Returns information about a variant.
	 *
	 * @param variant - Filter variant
	 */
	static describeVariant(variant: number): VariantDescription;
	
	/**
	 * Prints help.
	 */
	static printHelp(): void;
	
	/**
	 * @param variant - Filter variant
	 */
	constructor(variant: number);
	
	/**
	 * Adds an element to the set.
	 *
	 * @param uuid - UUID
	 */
	add(uuid: string): void;
	
	/**
	 * Tests whether an element is in the set.
	 *
	 * @param uuid - UUID
	 */
	has(uuid: string): boolean;
	
}


src/index.js
'use strict';

const {constants: {MAX_LENGTH: MAX_BUFFER_SIZE}} = require('buffer');
const {parse: uuidParse}                         = require('uuid');

class BloomFilter {
	
	buffers;
	variant;
	
	static describeVariant(variant) {
		if (!variant || (variant < 2)) throw new TypeError('Variant should be >= 2');
		
		const k = Math.floor(128 / variant);
		const m = k * (2**variant);
		
		const calcE = n => (1 - Math.exp(-k * n / m))**k;
		const calcN = e => Math.round(-m * Math.log(1 - e**(1/k)) / k);
		
		return {k, m, bufferSize: m / 8, calcE, calcN};
	}
	
	static printHelp() {
		const errors = [0.001, 0.01, 0.1, 1, 3, 5, 10, 15, 20, 25, 50];
		
		const help = [
			'A Bloom filter is a space-efficient probabilistic data structure, that is used to test whether an element is a member of a set.',
			'False positive matches are possible, but false negatives are not – in other words, a query returns either "possibly in set" or "definitely not in set".',
			'Elements can be added to the set, but not removed; the more items added, the larger the probability of false positives.\n'
		];
		
		const joiner     = ' | ';
		const columns    = ['Variant', centerText('Memory', 10), ...errors.map(n => centerText(n + '%', 8))];
		const tableTitle = columns.join(joiner)
		help.push(centerText('Max. number of inserted elements for desired false positive probability', tableTitle.length));
		help.push(tableTitle);
		help.push(columns.map(s => ''.padStart(s.length, '-')).join('-+-'));
		
		for (let i = 40; i > 1; i--) {
			const {bufferSize, calcN} = BloomFilter.describeVariant(i);
			help.push([
				centerText(String(i).padStart(2, ' '), columns[0].length),
				prettySize(bufferSize),
				...errors.map(e => prettyNumber(calcN(e / 100)))
			].join(joiner));
		}
		
		console.log(help.join('\n') + '\n');
		
		
		
		function centerText(text, columnWidth) {
			text = String(text);
			const paddingLeft = Math.ceil((columnWidth - text.length) / 2);
			return text.padStart(text.length + paddingLeft, ' ').padEnd(columnWidth, ' ');
		}
		
		function prettyNumber(value) {
			const n = Number(value.toPrecision(3));
			
			let divider = 1;
			let suffix  = ' ';
			for (const [d, s] of [[1e12, 'T'], [1e9, 'B'], [1e6, 'M'], [1e3, 'K']]) {
				if (n < d) continue;
				divider = d;
				suffix  = s;
				break;
			}
			
			const [ip, fp] = String(n / divider).split('.');
			return ip.padStart(3, ' ') + (fp ? `.${fp}` : '').padEnd(4, ' ') + suffix;
		}
		
		function prettySize(value) {
			let divider = 1;
			let suffix  = 'B';
			for (const [d, s] of [[2**40, 'TiB'], [2**30, 'GiB'], [2**20, 'MiB'], [2**10, 'KiB']]) {
				if (value < d) continue;
				divider = d;
				suffix  = s;
				break;
			}
			
			const [ip, fp] = String(value / divider).split('.');
			return ip.padStart(3, ' ') + (fp ? `.${fp}` : '').padEnd(4, ' ') + suffix.padStart(3, ' ');
		}
	}
	
	constructor(variant) {
		this.variant = variant;
		
		const {bufferSize} = BloomFilter.describeVariant(variant);
		const buffersSizes = [];
		{
			let leftSize = bufferSize;
			while (leftSize) {
				const length = Math.min(leftSize, MAX_BUFFER_SIZE);
				buffersSizes.push(length);
				leftSize -= length;
			}
		}
		this.buffers = buffersSizes.map(size => Buffer.alloc(size));
	}
	
	add(uuid) {
		const indexes = this._getIndexes(uuid);
		
		for (const index of indexes) {
			const bufferIndex = Math.floor(Math.floor(index / 8) / MAX_BUFFER_SIZE);
			const byteIndex   = Math.floor(index / 8) % MAX_BUFFER_SIZE;
			const bitIndex    = index % 8;
			
			this.buffers[bufferIndex][byteIndex] |= 2**bitIndex;
		}
	}
	
	has(uuid) {
		const indexes = this._getIndexes(uuid);
		
		for (const index of indexes) {
			const bufferIndex = Math.floor(Math.floor(index / 8) / MAX_BUFFER_SIZE);
			const byteIndex   = Math.floor(index / 8) % MAX_BUFFER_SIZE;
			const bitIndex    = index % 8;
			
			if (!(this.buffers[bufferIndex][byteIndex] & (2**bitIndex))) return false;
		}
		
		return true;
	}
	
	_getIndexes(uuid) {
		const uuidBuffer = Buffer.from(uuidParse(uuid));
		
		// Replace UUID version and variant
		uuidBuffer[6] = (uuidBuffer[6] & 0xf)
				+ ((uuidBuffer[0] & 1) << 4)
				+ ((uuidBuffer[1] & 1) << 5)
				+ ((uuidBuffer[2] & 1) << 6)
				+ ((uuidBuffer[3] & 1) << 7);
		uuidBuffer[8] = (uuidBuffer[8] & 0o77)
				+ ((uuidBuffer[4] & 1) << 6)
				+ ((uuidBuffer[5] & 1) << 7);
		
		const k = Math.floor(128 / this.variant);
		
		const groups     = [];
		let   byteN      = 0;
		let   leftInByte = 8;
		for (let i = 0; i < k; i++) {
			const group       = [];
			let   leftInGroup = this.variant;
			while (leftInGroup) {
				if (!leftInByte) {
					byteN++;
					leftInByte = 8;
				}
				
				const maskSize  = Math.min(leftInGroup, leftInByte);
				const maskShift = leftInByte  - maskSize;
				const bitsShift = leftInGroup - maskSize;
				group.push([byteN, maskSize, maskShift, bitsShift]);
				
				leftInGroup -= maskSize;
				leftInByte  -= maskSize;
			}
			groups.push(group);
		}
		
		const arr = [];
		for (const group of groups) {
			let n = 0;
			for (const [byteN, maskSize, maskShift, bitsShift] of group) {
				n += ((uuidBuffer[byteN] & (((2**maskSize) - 1) << maskShift)) >> maskShift) * (2**bitsShift);
			}
			arr.push(n);
		}
		
		return arr.map((n, i) => n + i * (2**this.variant));
	}
	
}

module.exports = BloomFilter;
